### DISCUSSION TITLE
error: only functions or types with a copy constructor can be applied

### DISCUSSION THREAD

**Discussion by chtenb** [koka-lang/koka#434] [Q&A]
Example code that demonstrates this error:

```koka
fun copy-and-replace-at(v : vector<a>, index : int, new : a) : exn vector<a>
  var result := v
  result[index] := new
  result
```

Full error:

```
repro.kk(3, 9): error: only functions or types with a copy constructor can be applied
 context      :         [index] := new
 term         :         [index]
 inferred type: int
```

What does this error mean exactly, and how would I write code like this without triggering this error?

**Comment by anfelor** [comment]
This looks like a bug connected to the `index` variable name. The following type-checks:

```koka
fun copy-and-replace-at(v : vector<a>, idx : int, new : a) : <div,exn> vector<a>
  var result := v
  result[idx] := new
  result
```

**Reply by anfelor** [reply]
In particular, I could imagine that this is an overloading error since the `index` function of `cctx` in `core/types` is not qualified? See #433

**Reply by chtenb** [reply]
For a moment I thought you meant the missing `div` effect was the culprit, but it's the name of the variable :) The `div` effect can be removed I think

**Reply by chtenb** [reply]
Hm, interesting. The div can only be removed if the vector type is monomorphic, like `vector<int>`.
Should I create an issue for this `index` name problem?

**Reply by anfelor** [reply]
The `div` effect is there in case the vector contains functions -- in that case we could write a function closure `new` into the vector, where `new` captures the vector itself. This way it is possible to create non-terminating programs: `new` simply has to call the function at position `v[idx]` which is just `new` again.

**Reply by anfelor** [reply]
I think this is the same issue as #433, but maybe @daanx or @TimWhiting disagree?

**Reply by TimWhiting** [reply]
No, I don't think it is the same issue. But I do think it is an issue that we should fix. I created an issue linking to this discussion.

**Comment by daanx** [comment]
Ah this was just a bug introduced when creating more consistent naming and (hidden) identifiers. It is fixed now in `dev`.   That the `div` effect occurs only on polymorphic vectors is really interesting :-)  Generally, this could be used to create divergence without having explicit recursion through "Landin's knot" construction.  I thought I had a section in the online book but I can't find it now ... I don't have time now to explain in depth but look at the documentation for `hdiv` (if I find time later I'll show the example).

**Reply by chtenb** [reply]
Thanks for the reply! I think your referring to this paper https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/paper-20.pdf
